ksmt-core/src/main/kotlin/org/ksmt/solver/model/KModelEvaluator.kt

```diff
@@ -283,97 +291,208 @@ open class KModelEvaluator(
                 "${interpretation.vars.size} arguments expected but ${args.size} provided"
             }
 
-            val resolvedInterpretation = ctx.resolveFunctionInterpretationComplete(interpretation, args)
-
-            // Interpretation was fully resolved
-            if (resolvedInterpretation.entries.isEmpty() && resolvedInterpretation.default != null) {
-                return@getOrPut resolvedInterpretation.default
+            val resolvedInterpretation = resolvedFunctionInterpretations.getOrPut(interpretation) {
+                resolveFunctionInterpretation(interpretation)
             }
 
-            // Interpretation was not fully resolved --> generate ITE chain
-            evalResolvedFunctionInterpretation(resolvedInterpretation, args)
+            ctx.applyResolvedInterpretation(resolvedInterpretation, args)
         }
         return evaluated.asExpr(decl.sort)
     }
 
-    private fun <T : KSort> evalResolvedFunctionInterpretation(
-        resolvedInterpretation: KModel.KFuncInterp<T>,
+    private fun <T : KSort> completeModelValue(sort: T): KExpr<T> {
+        val value = when (sort) {
+            is KUninterpretedSort ->
+                model.uninterpretedSortUniverse(sort)
+                    ?.randomOrNull()
+                    ?: sort.sampleValue()
+
+            is KArraySortBase<*> -> {
+                val arrayValue = completeModelValue(sort.range)
+                ctx.mkArrayConst(sort.uncheckedCast(), arrayValue)
+            }
+
+            else -> sort.sampleValue()
+        }
+        return value.asExpr(sort)
+    }
+
+    private fun <T : KSort> resolveFunctionInterpretation(
+        interpretation: KModel.KFuncInterp<T>
+    ): ResolvedFunctionInterpretation<T> {
+        var resolvedEntry: ResolvedFunctionEntry<T> = ResolvedFunctionDefaultEntry(interpretation.default)
+
+        for (entry in interpretation.entries.asReversed()) {
+            val isValueEntry = entry.args.all { isValueInModel(it) }
+
+            resolvedEntry = if (isValueEntry) {
+                resolvedEntry.addValueEntry(entry.args, entry.value)
+            } else {
+                resolvedEntry.addUninterpretedEntry(entry.args, entry.value)
+            }
+        }
+
+        return ResolvedFunctionInterpretation(interpretation, resolvedEntry)
+    }
+
+    private fun <T : KSort> KContext.applyResolvedInterpretation(
+        interpretation: ResolvedFunctionInterpretation<T>,
         args: List<KExpr<*>>
+    ): KExpr<T> {
+        val argsAreConstants = args.all { isValueInModel(it) }
```

Comment:
Can be moved into tge `ResolvedFunctionValuesEntry` branch

Reply:
We don't want to recompute this value on each iteration.